cleaner.go

```diff
@@ -0,0 +1,362 @@
+	cleaner.docsToProcess = cleaner.docsToProcess[:0]
+	return nil
+}
+
+// flushRemoveQueue ensures that all pending removals are flushed to the database.
+func (cleaner *collectionCleaner) flushRemoveQueue() error {
+	if len(cleaner.docIdsToRemove) == 0 {
+		return nil
+	}
+	remover := newBulkRemover(cleaner.config.Source)
+	for _, docId := range cleaner.docIdsToRemove {
+		if err := remover.remove(docId); err != nil {
+			return fmt.Errorf("failed while removing document %v from %q", docId, cleaner.config.Source.Name)
+		}
+	}
+	if err := remover.flush(); err != nil {
+		return fmt.Errorf("failed while removing documents from %q", cleaner.config.Source.Name)
+	}
+	cleaner.stats.RemovedCount += remover.removed
+	logger.Debugf("flushing %d documents removed %d (%d total)",
+		len(cleaner.docIdsToRemove), remover.removed, cleaner.stats.RemovedCount)
+	cleaner.docIdsToRemove = cleaner.docIdsToRemove[:0]
+	return nil
+}
+
+// Cleanup iterates the collection and ensures that all documents no longer reference completed transactions.
+func (cleaner *collectionCleaner) Cleanup() error {
+	// walk through all the documents and remove what we can, update what we cannot
+	logger.Debugf("cleaning up completed documents from %q", cleaner.config.Source.Name)
+	// If we delete documents while we iterate, it can cause the iterator to miss documents.
+	// So we do multiple passes on the database to make sure we catch everything.
+	var doc txnDocument
+	for iterCount := 0; iterCount < maxIterCount; iterCount++ {
+		removedWhileIterating := false
+		query := cleaner.config.Source.Find(nil)
+		query.Batch(maxBatchDocs)
+		iter := query.Iter()
+		for iter.Next(&doc) {
+			if cleaner.massageId != nil {
+				doc.Id = cleaner.massageId(doc.Id)
+			}
+			if shouldFlush, err := cleaner.includeDoc(doc); err != nil {
+				return err
+			} else if shouldFlush {
+				if err := cleaner.processStashDocs(); err != nil {
+					return err
+				}
+				if len(cleaner.docIdsToRemove) > cleaner.config.MaxRemoveQueue {
+					if err := cleaner.flushRemoveQueue(); err != nil {
+						return err
+					}
+					removedWhileIterating = true
+				}
+			}
+		}
+		if err := iter.Close(); err != nil {
+			return fmt.Errorf("error while iterating %q: %v", cleaner.config.Source.Name, err)
+		}
+		if err := cleaner.processStashDocs(); err != nil {
+			return err
+		}
+		if err := cleaner.flushRemoveQueue(); err != nil {
+			return err
+		}
+		if !removedWhileIterating {
+			break
+		}
+	}
+	logger.Debugf("%s", cleaner.stats.Details())
+	if cleaner.removeIfEmpty {
+		finalCount, _ := cleaner.config.Source.Count()
+		logger.Debugf("%s has %d documents left", cleaner.config.Source.Name, finalCount)
+	}
+	return nil
+}
+
+// CleanupStash goes through the txns.stash and removes documents that are no longer needed.
+func CleanupStash(db *mgo.Database, txns, txnsStash *mgo.Collection) error {
+	batchTokens := queueBatchSize
+	if size := os.Getenv("JUJU_QUEUE_BATCH_SIZE"); size != "" {
```

Comment:
Do these env vars really need to stay? If so, can you make the names a little less generic? They should at least mention "txn" and "prune" IMO.